File: src/main/java/uk/laxd/dndSimulator/dice/Die.java
package uk.laxd.dndSimulator.dice;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Random;

/**
 * Represents a single die, which can be rolled to get a random number
 * within its range
 */
public class Die {

    private static final Logger LOGGER = LoggerFactory.getLogger(Die.class);

    public static final Die D4 = new Die(4);
    public static final Die D6 = new Die(6);
    public static final Die D8 = new Die(8);
    public static final Die D12 = new Die(12);
    public static final Die D20 = new Die(20);

    private int sides;

    public Die(int sides) {
        this.sides = sides;
    }

    public int roll() {
        int result = new Random().nextInt(sides) + 1;
        LOGGER.debug("Rolled a {} = {}", this, result);
        return result;
    }

    public int getMaxValue() {
        return sides;
    }

    @Override
    public String toString() {
        return String.format("d%s", sides);
    }
}

File: app/src/main/kotlin/uk/laxd/dndSimulator/dice/Die.kt
package uk.laxd.dndSimulator.dice

import org.slf4j.LoggerFactory
import java.util.*
import kotlin.math.max
import kotlin.math.min

/**
 * Represents a single die, which can be rolled to get a random number
 * within its range
 */
open class Die(val maxValue: Int) {
    private val logger = LoggerFactory.getLogger(Die::class.java)

    open fun roll(): Int {
        val result = Random().nextInt(maxValue) + 1
        logger.debug("Rolled a {} = {}", this, result)
        return result
    }

    override fun toString(): String {
        return "d$maxValue"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Die) return false

        if (maxValue != other.maxValue) return false

        return true
    }

    override fun hashCode(): Int {
        return maxValue
    }

    companion object {
        val D4 = Die(4)
        @JvmField
        val D6 = Die(6)
        @JvmField
        val D8 = Die(8)
        @JvmField
        val D10 = Die(10)
        @JvmField
        val D12 = Die(12)
        @JvmField
        val D20 = Die(20)

        @JvmField
        val D20_Advantage = AdvantageDie(D20)

        @JvmField
        val D20_Disadvantage = DisadvantageDie(D20)

        fun fixedRoll(value: Int, maxValue: Int): Die {
            return FixedDie(value, maxValue)
        }
    }
}

class AdvantageDie(val die: Die): Die(die.maxValue) {
    private val logger = LoggerFactory.getLogger(AdvantageDie::class.java)
    override fun roll(): Int {
        val die1 = die.roll()
        val die2 = die.roll()

        val result = max(die1, die2)

        logger.debug("Rolled a {} with advantage = {} and {} = {}", this, die1, die2, result)

        return result
    }
}

class DisadvantageDie(val die: Die): Die(die.maxValue) {
    private val logger = LoggerFactory.getLogger(DisadvantageDie::class.java)
    override fun roll(): Int {
        val die1 = die.roll()
        val die2 = die.roll()

        val result = min(die1, die2)

        logger.debug("Rolled a {} with disadvantage = {} and {} = {}", this, die1, die2, result)

        return result
    }
}